File: CifrarMensagem.cpp
#include <iostream>
#include <algorithm>
#include <random>
#include <string>
using namespace std;

void geraTabelaPadrao();
void geraTabelaAleatoria();
void mostraMenuDeOpcoes();
void imprimeTabela();
string cifraMensagem(string mensagemDescriptografada);
string preparaMensagem(string mensagem);
string removeEspacos(string mensagem);
bool estaNaMesmaLinha(char primeiraLetra, char segundaLetra);
bool estaNaMesmaColuna(char primeiraLetra, char segundaLetra);
char letraDaDireita(char letra);
char letraDeBaixo(char letra);
char correspondente(char primeiraLetra, char segundaLetra);
int getLinhaNaMatriz(char letra);
int getColunaNaMatriz(char letra);

const int NOVA_CIFRA = 1;
const int CIFRAR_MENSAGEM = 2;
const int VER_MENSAGEM_CIFRADA = 3;
const int VER_MENSAGEM_ORIGINAL = 4;
const int VER_ALFABETO = 5;
const int TERMINAR = 6;
const int TAMANHO_MATRIZ = 5;
const string ALFABETO ("ABCDEFGHIJKLMNOPQRSTUVWXZ");

char tabela[5][5];

int main() {
	geraTabelaPadrao();
	
	string mensagemDecifrada ("Voce ainda nao introduziu uma mensagem para cifrar.");
	string mensagemCifrada ("Voce ainda nao introduziu uma mensagem para cifrar.");
	string dummy;

	int opcao = -1;

	while (opcao != TERMINAR) {
		mostraMenuDeOpcoes();
		
		cin >> opcao;
		getline(cin, dummy);
		cout << "\n";

		switch(opcao) {
			case NOVA_CIFRA:
				geraTabelaAleatoria();
				cout << "A tabela da cifra foi atualizada com sucesso!\n";
				break;
			case CIFRAR_MENSAGEM:
				cout << "Digite a mensagem que sera cifrada: ";
				getline(cin, mensagemDecifrada);
				mensagemCifrada = cifraMensagem(mensagemDecifrada);
				cout << "Mensagem cifrada com sucesso!\n";
				break;
			case VER_MENSAGEM_CIFRADA:
				cout << "Essa eh a sua mensagem cifrada: " << mensagemCifrada << "\n";
				break;
			case VER_MENSAGEM_ORIGINAL:
				cout << "Essa eh a sua mensagem original: " << mensagemDecifrada << "\n";
				break;
			case VER_ALFABETO:
				cout << "Tabela da cifra:\n";
				imprimeTabela();
				break;
			case TERMINAR:
				break;
			default:
				mostraMenuDeOpcoes();
				break;
		}
		
		cout << "\n";
	}
}

void geraTabelaPadrao() {
	tabela[0][0] = 'Y'; tabela[0][1] = 'Q'; tabela[0][2] = 'D'; tabela[0][3] = 'L'; tabela[0][4] = 'G';
	tabela[1][0] = 'M'; tabela[1][1] = 'J'; tabela[1][2] = 'X'; tabela[1][3] = 'F'; tabela[1][4] = 'U';
	tabela[2][0] = 'V'; tabela[2][1] = 'W'; tabela[2][2] = 'C'; tabela[2][3] = 'P'; tabela[2][4] = 'B';
	tabela[3][0] = 'O'; tabela[3][1] = 'S'; tabela[3][2] = 'K'; tabela[3][3] = 'R'; tabela[3][4] = 'E';
	tabela[4][0] = 'T'; tabela[4][1] = 'H'; tabela[4][2] = 'N'; tabela[4][3] = 'A'; tabela[4][4] = 'I';
}

void geraTabelaAleatoria() {
	string alfabeto (ALFABETO);
	
	shuffle (alfabeto.begin(), alfabeto.end(), random_device());
	
	string::iterator it = alfabeto.begin();
	int i, j;
	for (i = 0; i < TAMANHO_MATRIZ; i++) {
		for (j = 0; j < TAMANHO_MATRIZ; j++) {
			tabela[i][j] = *it++;
		}
	}
}

void mostraMenuDeOpcoes() {
	cout << "Escolha uma das opcoes abaixo:\n\n";
	cout << "1. Escolher uma tabela de cifra nova\n";
	cout << "2. Introduzir uma mensagem para cifrar\n";
	cout << "3. Ver a mensagem cifrada\n";
	cout << "4. Decifrar a mensagem\n";
	cout << "5. Ver o alfabeto\n";
	cout << "6. Terminar\n\n";
	cout << "Opcao: ";
}

void imprimeTabela() {
	int i, j;
	for (i = 0; i < TAMANHO_MATRIZ; i++) {
		for (j = 0; j < TAMANHO_MATRIZ; j++) {
			cout << tabela[i][j] << " ";
		}
		cout << "\n";
	}
}

string cifraMensagem(string mensagemDescriptografada) {
		string mensagemCriptografada = (""), mensagemPreparada = preparaMensagem(mensagemDescriptografada);
		char primeiraLetra, segundaLetra;

		bool juntaPares = true;
		
		for (int i = 0; i < mensagemPreparada.length(); i+=3) {
			primeiraLetra = mensagemPreparada[i];
			segundaLetra = mensagemPreparada[i+1];
			
			if (estaNaMesmaLinha(primeiraLetra, segundaLetra)) {
				mensagemCriptografada += letraDaDireita(primeiraLetra);
				mensagemCriptografada += letraDaDireita(segundaLetra);
			} else if (estaNaMesmaColuna(primeiraLetra, segundaLetra)) {
				mensagemCriptografada += letraDeBaixo(primeiraLetra);
				mensagemCriptografada += letraDeBaixo(segundaLetra);
			} else {
				mensagemCriptografada += correspondente(primeiraLetra, segundaLetra);
				mensagemCriptografada += correspondente(segundaLetra, primeiraLetra);
			}
			
			mensagemCriptografada += (juntaPares ? "" : " ");
			juntaPares = !juntaPares;
		}
	
		return mensagemCriptografada;
	}

string preparaMensagem(string mensagem) {
	string mensagemPreparada = ("");
	
	mensagem = removeEspacos(mensagem);

	for (int i = 0; i < mensagem.length(); i++) {
		mensagemPreparada += mensagem[i];
		
		if (i == (mensagem.length()-1) || mensagem[i] == mensagem[i+1]) {
			mensagemPreparada += "X ";
		} else {
			mensagemPreparada += mensagem[i+1];
			mensagemPreparada += " ";
			i++;
		}
	}
	
	return mensagemPreparada;
}

string removeEspacos(string mensagem) {
	string mensagemSemEspaco ("");
	
	for(int i = 0; i < mensagem.length(); i++) {
		if(mensagem[i] != ' ') {	
			mensagemSemEspaco += mensagem[i]; 
		}
	}
	
	return mensagemSemEspaco;
}

bool estaNaMesmaLinha(char primeiraLetra, char segundaLetra) {
	int linha = getLinhaNaMatriz(primeiraLetra);
	
	int coluna = 0;
	
	while (coluna < TAMANHO_MATRIZ) {
		if (tabela[linha][coluna] == segundaLetra) {
			return true;
		}
		
		coluna++;
	}

	return false;
}

bool estaNaMesmaColuna(char primeiraLetra, char segundaLetra) {
	int coluna = getColunaNaMatriz(primeiraLetra);
	
	int linha = 0;
	
	while (linha < TAMANHO_MATRIZ) {
		if (tabela[linha][coluna] == segundaLetra) {
			return true;
		}
		
		linha++;
	}

	return false;
}


char letraDaDireita(char letra) {
	int linha = getLinhaNaMatriz(letra);
	int coluna = getColunaNaMatriz(letra);

	return tabela[linha][(coluna + 1) % TAMANHO_MATRIZ];
}

char letraDeBaixo(char letra) {
	int linha = getLinhaNaMatriz(letra);
	int coluna = getColunaNaMatriz(letra);
	
	return tabela[(linha + 1) % TAMANHO_MATRIZ][coluna];
}

char correspondente(char primeiraLetra, char segundaLetra) {
	int linha = getLinhaNaMatriz(primeiraLetra);
	int coluna = getColunaNaMatriz(segundaLetra);
	
	return tabela[linha][coluna];
}

int getLinhaNaMatriz(char letra) {
	bool encontrada = false;
	int linha = 0, coluna = 0;
	
	while (!encontrada && linha < TAMANHO_MATRIZ) {
		while (!encontrada && coluna < TAMANHO_MATRIZ) {
			if (tabela[linha][coluna] == letra) {
				encontrada = true;
			}
			if (!encontrada) coluna++;
		}
		if (!encontrada) {
			linha++;
			coluna = 0;
		}
	}
	
	return linha;
}

int getColunaNaMatriz(char letra) {
	bool encontrada = false;
	int linha = 0, coluna = 0;
	
	while (!encontrada && linha < TAMANHO_MATRIZ) {
		while (!encontrada && coluna < TAMANHO_MATRIZ) {
			if (tabela[linha][coluna] == letra) {
				encontrada = true;
			}
			if (!encontrada) coluna++;
		}
		if (!encontrada) {
			linha++;
			coluna = 0;
		}
	}
	
	return coluna;
}

File: CifrarMensagem.hs
import Data.Char

_CIFRAR_MENSAGEM = 1
_INDICE_TABELA = 4
_TAMANHO_TABELA = 5
_TABELA = ([['Y','Q','D','L','G'],
			['M','J','X','F','U'],
			['V','W','C','P','B'],
			['O','S','K','R','E'],
			['T','H','N','A','I']])

main :: IO ()
main = do
  mostrarMenuDeOpcoes
  
  opcaoArgs <- getLine
  let opcao = read opcaoArgs
  
  mensagemArgs <- getLine
  let mensagem = read mensagemArgs
  
  let mensagemPreparada = preparaMensagem mensagem
  let resultadoProcessamento = processa opcao mensagemPreparada
  
  putStrLn ("\n")
  
  if not (ehVazia resultadoProcessamento) then do
	{ putStrLn (resultadoProcessamento ++ "\n"); main }
  else do
    { putStrLn ("Volte sempre! :)") }

preparaMensagem :: String -> String
preparaMensagem frase = juntaDoisADois fraseValida
    where fraseValida = removeCaracInvalidos frase

juntaDoisADois :: String -> String
juntaDoisADois [] = ""
juntaDoisADois (x:[]) = (x:'X':[])
juntaDoisADois (x:y:xs) = if x == y then (x:'X':[]) ++ juntaDoisADois (y:xs)
						  else(x:y:[]) ++ juntaDoisADois xs

removeCaracInvalidos :: String -> String
removeCaracInvalidos frase = [ x | x <- frase, isLetter x ]

mostrarMenuDeOpcoes :: IO()
mostrarMenuDeOpcoes = do
  putStrLn ("Escolha uma das opcoes abaixo:\n")
  putStrLn ("1. Cifrar mensagem")
  putStrLn ("2. Ver o alfabeto")
  putStrLn ("3. Terminar\n")
  putStrLn ("Opcao: ")

processa :: Int -> String -> String
processa 1 mensagem = adicionaEspacoCadaQuatro (cifraMensagem mensagem)
processa 2 mensagem = transformaTabela _TABELA
processa opcaoInvalida mensagem = ""

transformaTabela :: [[Char]] -> String
transformaTabela [] = ""
transformaTabela (x:xs) = adicionaEspacoEntreCaracters x ++ "\n" ++ transformaTabela xs

adicionaEspacoEntreCaracters :: String -> String
adicionaEspacoEntreCaracters [] = ""
adicionaEspacoEntreCaracters (x:xs) = x:' ':[] ++ adicionaEspacoEntreCaracters xs

ehVazia :: String -> Bool
ehVazia "" = True
ehVazia string = False

estaNaMesmaLinha :: Char -> Char -> Bool
estaNaMesmaLinha primeiraLetra segundaLetra = indiceLinhaPrimeiraLetra == indiceLinhaSegundaLetra
    where indiceLinhaPrimeiraLetra = getLinhaNaMatriz primeiraLetra _INDICE_TABELA
          indiceLinhaSegundaLetra  = getLinhaNaMatriz segundaLetra _INDICE_TABELA

getLinhaNaMatriz :: Char -> Int -> Int
getLinhaNaMatriz letra 0 = if letra `elem` (_TABELA !! 0) then 0 else -1
getLinhaNaMatriz letra indiceLinha = if letra `elem` linhaAtual then indiceLinha 
                                     else getLinhaNaMatriz letra (indiceLinha-1)
    where linhaAtual = _TABELA !! indiceLinha

estaNaMesmaColuna :: Char -> Char -> Bool
estaNaMesmaColuna primeiraLetra segundaLetra = indiceColunaPrimeiraLetra == indiceColunaSegundaLetra
    where indiceColunaPrimeiraLetra = getColunaNaMatriz primeiraLetra (getLinhaNaMatriz primeiraLetra _INDICE_TABELA) _INDICE_TABELA
          indiceColunaSegundaLetra  = getColunaNaMatriz segundaLetra (getLinhaNaMatriz segundaLetra _INDICE_TABELA) _INDICE_TABELA

getColunaNaMatriz :: Char -> Int -> Int -> Int
getColunaNaMatriz letra indiceLinha 0 = if letra == _TABELA !! indiceLinha !! 0 then 0 else -1
getColunaNaMatriz letra indiceLinha indiceColuna = if letra == letraDaColunaAtual then indiceColuna 
												   else getColunaNaMatriz letra indiceLinha (indiceColuna-1)
    where letraDaColunaAtual = _TABELA !! indiceLinha !! indiceColuna

letraDaDireita :: Char -> Char
letraDaDireita letra = let linha = getLinhaNaMatriz letra _INDICE_TABELA in let coluna = getColunaNaMatriz letra linha _INDICE_TABELA in
                              _TABELA !! linha !! ((coluna+1) `mod` _TAMANHO_TABELA)

letraDeBaixo :: Char -> Char
letraDeBaixo letra = let linha = getLinhaNaMatriz letra _INDICE_TABELA in let coluna = getColunaNaMatriz letra linha _INDICE_TABELA in
                            _TABELA !! ((linha+1) `mod` _TAMANHO_TABELA) !! coluna

correspondente :: Char -> Char -> Char
correspondente primeiraLetra segundaLetra = let linha = getLinhaNaMatriz primeiraLetra _INDICE_TABELA in 
                                                   let coluna = getColunaNaMatriz segundaLetra (getLinhaNaMatriz segundaLetra _INDICE_TABELA) _INDICE_TABELA in
                                                   _TABELA !! linha !! coluna

adicionaEspacoCadaQuatro :: String -> String
adicionaEspacoCadaQuatro frase = if length frase <= 4 then frase
							  else take 4 frase ++ " " ++ adicionaEspacoCadaQuatro (drop 4 frase)

cifraMensagem :: String -> String
cifraMensagem [] = []
cifraMensagem (a:b:cs) =  if estaNaMesmaLinha a b then letraDaDireita a : letraDaDireita b : cifraMensagem cs
						  else if estaNaMesmaColuna a b then letraDeBaixo a : letraDeBaixo b : cifraMensagem cs
						  else correspondente a b : correspondente b a : cifraMensagem cs
